crates/sui-types/src/move_package.rs

```diff
@@ -129,10 +161,65 @@ impl MovePackage {
         Ok(pkg)
     }
 
-    pub fn from_module_iter<T: IntoIterator<Item = CompiledModule>>(
+    /// Create an initial version of the package along with this version's type origin and linkage
+    /// tables.
+    pub fn new_initial(
+        version: SequenceNumber,
+        modules: Vec<CompiledModule>,
+        max_move_package_size: u64,
+        dependencies: &[MovePackage],
+    ) -> Result<Self, ExecutionError> {
+        let type_origin_table = Self::build_initial_type_origin_table(&modules);
+        Self::from_module_iter_with_type_origin_table(
+            version,
+            modules,
+            max_move_package_size,
+            type_origin_table,
+            dependencies,
+        )
+    }
+
+    /// Create an upgraded version of the package along with this version's type origin and linkage
+    /// tables.
+    pub fn new_upgraded(
+        predecessor: &MovePackage,
+        modules: Vec<CompiledModule>,
+        max_move_package_size: u64,
+        dependencies: &[MovePackage],
+    ) -> Result<Self, ExecutionError> {
+        let type_origin_table = Self::build_upgraded_type_origin_table(predecessor, &modules);
+        let mut new_version = SequenceNumber::from(predecessor.version().value());
+        new_version.increment();
```

Comment:
nit: `SequenceNumber` has `Copy` so you don't need to use `from` and `value` to clone it.

```suggestion
        let mut new_version = predecessor.version();
        new_version.increment();
```